Lambda sort and DatatypeRef in _to_expr_ref

I'm looking at getting cvc5 to be an alternate backend to my project and there are a couple incompatibilities.

There's a missing cast to DatatypeRef here https://github.com/cvc5/cvc5_pythonic_api/blob/be54c2388b3271f657cad41cf5e3d6bc97cd51a1/cvc5_pythonic_api/cvc5_pythonic.py#L949

which I backpatched as

```
def Const(name, sort):
    # _to_expr doesn't have a DatatypeRef case
    x = cvc5.pythonic.Const(name, sort)
    if isinstance(sort, DatatypeSortRef):
        x = DatatypeRef(x.ast, x.ctx, x.reverse_children)
    return x
```

And the sort returned by lambdas is not right https://github.com/cvc5/cvc5_pythonic_api/blob/be54c2388b3271f657cad41cf5e3d6bc97cd51a1/cvc5_pythonic_api/cvc5_pythonic.py#L9031

I tried patching this as 
```
def _qsort(self):
    if self.is_lambda():
        return ArraySort(self.var_sort(0), self.body().sort())
    else:
        return BoolSort(self.ctx)


QuantifierRef.sort = _qsort

```
but I get a deeper error which I don't understand

```
________________________________________________________________ ERROR collecting tests/test_kernel.py ________________________________________________________________
tests/test_kernel.py:9: in <module>
    import knuckledragger.theories.Interval
knuckledragger/theories/Interval.py:9: in <module>
    setof = kd.define("setof", [i], smt.Lambda([x], smt.And(i.lo <= x, x <= i.hi)))
knuckledragger/kernel.py:132: in define
    f(*args) == body
../../../.local/lib/python3.10/site-packages/cvc5/pythonic/cvc5_pythonic.py:375: in __eq__
    return BoolRef(c.solver.mkTerm(Kind.EQUAL, a.as_ast(), b.as_ast()), c)
cvc5.pxi:2556: in cvc5.cvc5_python_base.Solver.mkTerm
    ???
cvc5.pxi:1538: in cvc5.cvc5_python_base.TermManager.mkTerm
    ???
E   RuntimeError: Subexpressions must have the same type:
E   Equation: (= (setof i) (lambda ((x Real)) (and (<= (lo i) x) (<= x (hi i)))))
E   Type 1: (Array Real Bool)
E   Type 2: (-> Real Bool)
```

Are lambda sorts distinct from Arrays?

--- a/cvc5_pythonic_api/cvc5_pythonic.py
+++ b/cvc5_pythonic_api/cvc5_pythonic.py
@@ -795,6 +795,8 @@ def _to_sort_ref(s, ctx):
         return FPRMSortRef(s, ctx)
     elif s.isFunction():
         return FuncSortRef(s, ctx)
+    elif s.isDatatype():
+        return DatatypeSortRef(s, ctx)
     return SortRef(s, ctx)
 
 
@@ -1088,6 +1090,8 @@ def _to_expr_ref(a, ctx, r=None):
         return SeqRef(ast, ctx, r)
     if sort.isFunction():
         return FuncDeclRef(ast, ctx, r)
+    if sort.isDatatype():
+        return DatatypeRef(ast, ctx, r)
     return ExprRef(ast, ctx, r)
 
 
@@ -9054,11 +9058,36 @@ class DatatypeRecognizerRef(FuncDeclRef):
 
 
 class DatatypeRef(ExprRef):
-    """Datatype expressions."""
+    """Datatype expressions.
+
+    Constants and applications of a datatype sort are instances of this class.
+
+    >>> List = Datatype('List')
+    >>> List.declare('cons', ('car', IntSort()), ('cdr', List))
+    >>> List.declare('nil')
+    >>> List = List.create()
+    >>> isinstance(Const('l', List), DatatypeRef)
+    True
+    >>> isinstance(List.cons(10, List.nil), DatatypeRef)
+    True
+    """
 
     def sort(self):
-        """Return the datatype sort of the datatype expression `self`."""
-        return DatatypeSortRef(self.as_ast().getSort(), self.ctx)
+        """Return the datatype sort of the datatype expression `self`.
+
+        >>> List = Datatype('List')
+        >>> List.declare('cons', ('car', IntSort()), ('cdr', List))
+        >>> List.declare('nil')
+        >>> List = List.create()
+        >>> l = Const('l', List)
+        >>> l.sort()
+        List
+        >>> l.sort().num_constructors()
+        2
+        >>> l.sort().accessor(0, 0)
+        car
+        """
+        return _sort(self.ctx, self.ast)
 
 
 def TupleSort(name, sorts, ctx=None):

